### PR TITLE
Selection of favorite teams and sort on the dashboard

### LINKED ISSUE #60: [dashboard] Team as favorite
It can be tricky to swipe all teams (especially for club which has a lot of teams) to display your favorite(s) one(s).
Now teams are sorted by their name, but favorite's teams must be displayed first.

   

- [x] Add the ability to set one or more teams as favorite (in the club page)
- [x] Store this favorite teams in the local storage as we've done for club (with a different key per club?). 
- [x] In the dashboard, display first all favorite teams and then other teams (but keep alphabetic order in each group)
- [x] Apply the same sort in the club page

### PR DESCRIPTION
**Pull request by thomasFoulon**
Close #60 

The user can select favorite teams in the club page. Once it's done, he can see favorite teams first on the dashboard. To indicate the user that a team is favorite on the dashboard, there are several solutions (see screenshots). The last commit implements the third solution, the one with the non clickable star which is personally my preferred solution.

What do you think ?

### CODE REVIEW

**Comment by eric-taix** [comment]
I have to checkout your branch to see how it is animated or could you record a gif animation and add it to this PR?
However first comments: why "Equipes favorites" is on the right and "Autres" on the left? Do these labels move with cards when it changes from a favorite to other? I know that I suggested you to use "Autres" but now I'm not sure it is explicit!

**Comment by eric-taix** [comment]
I'm wondering if the star is not the good solution to display that a team is a favorite one. WDYT ?

**Comment by thomasFoulon** [comment]
No the left and right side have nothing to do with "Equipes favorites" or "Autres", it's just two different solutions, one with the two on the left side and one with the two on the right side !

I personally prefer the black or white star on the dashboard's card to well indicate that it's not clickable.

I will try to do a GIF.

**Comment by eric-taix** [comment]
Ok so we are aligned. Let's go for the black/white star to indicate that a
team is a favorite one

Le ven. 10 juil. 2020 à 12:18, thomasFoulon <notifications@github.com> a
écrit :

**Comment by thomasFoulon** [comment]
That's the best I can do... I used `ffmpeg` to convert the `mp4` into a `gif` but I think I don't have correct parameters...

File: lib/pages/club/club_page.dart [hunk 1]
```diff
@@ -38,10 +38,16 @@ class _ClubPageState extends State<ClubPage> with SingleTickerProviderStateMixin
       _loading = true;
     });
     _repository.loadAllClubs().then((clubs) {
-      setState(() {
-        _loading = false;
-        clubs.sort((c1, c2) => c1.shortName.toUpperCase().compareTo(c2.shortName.toUpperCase()));
-        _clubs = clubs;
+      _repository.loadFavoriteClubCodes().then((favorites) {
```
Comment (eric-taix-teads) [inline lib/pages/club/club_page.dart:41]: 👍

**Comment by thomasFoulon** [comment]
Some bugs are staying. When the user selects favorite teams from clicking on the `FavoriteClubCard` on the dashboard, at some cases it works, then it doesn't.

**Comment by eric-taix-teads** [comment]
Ok I started to have a look and I found some mistakes (some have been introduced before this PR - sorry I didn't not spend enough time on previous reviews). I'll fix your PR and I'll comment all the difference, so you will be able to understand what was wrong.

**Comment by thomasFoulon** [comment]
Ok, thanks ! By the way, I'm sometimes lost in the code, especially in the dashboard. Can I create an issue to well separate elements in the dashboard (favorite clubs, favorite teams and agenda in 3 different widgets and not only in `dashboard_page.dart`) ?

**Comment by eric-taix** [comment]
Yes please do !

Le lun. 13 juil. 2020 à 13:30, thomasFoulon <notifications@github.com> a
écrit :

**Comment by eric-taix-teads** [comment]
@thomasFoulon I do not have the time to add some comments right now but you can have a look to the four commits I've added (some are not relevant to fix your PR). I don't want to spoil you but the main point is that you stored the state into the bloc: never do that, the state **must** be immutable and is materialized with an instance of `XXXState` (always) 

And if you have time, can you checkout this branch and verify that the bug has been solved?
There are some conflicts: I'll resolve them tonight.

File: lib/pages/dashboard/dashoard_page.dart [hunk 2]
```diff
@@ -119,9 +119,11 @@ class _DashboardPageState extends State<DashboardPage> {
                       physics: BouncingScrollPhysics(),
                       itemCount: state.clubs.length,
                       controller: _pageController,
-                      onPageChanged: (pageIndex) => _updateClubTeams(state.clubs[pageIndex].code),
```
Comment (eric-taix-teads) [inline lib/pages/dashboard/dashoard_page.dart:122]: This method only select the current club even if when selecting the current team, it reloads favorite club's teams

File: lib/pages/dashboard/fav_club_card.dart [hunk 3]
```diff
@@ -18,7 +18,7 @@ class FavoriteClubCard extends StatefulWidget {
   _FavoriteClubCardState createState() => _FavoriteClubCardState();
 }
 
-class _FavoriteClubCardState extends State<FavoriteClubCard> with AutomaticKeepAliveClientMixin {
```
Comment (eric-taix-teads) [inline lib/pages/dashboard/fav_club_card.dart:21]: Not sure it helps here?

**Comment by eric-taix-teads** [comment]
@thomasFoulon If you can have a look quickly and approve or comment if you see anything wrong.

**Pull request merged**